500 error on malformed filter

```
[
    {
        "@timestamp": "2025-09-10 16:44:54.109",
        "@message": "START RequestId: 9cba99a5-e83d-4a96-8369-0c550c1e5f12 Version: $LATEST\n"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.110",
        "@message": "[INFO]\t2025-09-10T16:44:54.110Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.chalice\tReceived GET request for '/index/files', with {\"query\": {\"filters\": \"{\\\"projectId\\\":{\\\"is\\\":[{}]},\\\"fileFormat\\\":{\\\"is\\\":[\\\"loom\\\"]}}\", \"order\": \"asc\", \"size\": \"2\", \"sort\": \"projectTitle\"}, \"headers\": {\"accept\": \"application/json, text/xml, application/xml, */*\", \"accept-encoding\": \"gzip\", \"cloudfront-forwarded-proto\": \"https\", \"cloudfront-is-desktop-viewer\": \"true\", \"cloudfront-is-mobile-viewer\": \"false\", \"cloudfront-is-smarttv-viewer\": \"false\", \"cloudfront-is-tablet-viewer\": \"false\", \"cloudfront-viewer-asn\": \"32934\", \"cloudfront-viewer-country\": \"US\", \"host\": \"[service.azul.data.humancellatlas.org](http://service.azul.data.humancellatlas.org/)\", \"user-agent\": \"facebookexternalhit/1.1 (+http://www.facebook.com/externalhit_uatext.php)\", \"via\": \"2.0 [4f8e8c9677a68ac8935c32689f962b7a.cloudfront.net](http://4f8e8c9677a68ac8935c32689f962b7a.cloudfront.net/) (CloudFront)\", \"x-amz-cf-id\": \"AR5IhKRk012Jr0PqvFNE5Sgxh5wc1BYpkkRIHyAJxqrN_XZWeTsAvA==\", \"x-amzn-trace-id\": \"Root=1-68c1ab06-742ec0161b7e2ad37e85a828\", \"x-forwarded-for\": \"57.141.4.70, 18.68.30.151\", \"x-forwarded-port\": \"443\", \"x-forwarded-proto\": \"https\"}}."
    },
    {
        "@timestamp": "2025-09-10 16:44:54.110",
        "@message": "[INFO]\t2025-09-10T16:44:54.110Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.chalice\t… without a request body"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.110",
        "@message": "[INFO]\t2025-09-10T16:44:54.110Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.chalice\tDid not authenticate request."
    },
    {
        "@timestamp": "2025-09-10 16:44:54.112",
        "@message": "[INFO]\t2025-09-10T16:44:54.111Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.boto3\tdynamodb.GetItem:\tMaking POST request to https://dynamodb.us-east-1.amazonaws.com/"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.112",
        "@message": "[INFO]\t2025-09-10T16:44:54.112Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.boto3\tdynamodb.GetItem:\t… with a request body of length 132 being b'{\"TableName\": \"azul-sources_cache_by_auth-prod\", \"Key\": {\"identity\": {\"S\": \"dcp52:\"}}, \"ProjectionExpression\": \"sources,expiration\"}'"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.118",
        "@message": "[INFO]\t2025-09-10T16:44:54.118Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.boto3\tdynamodb.GetItem:\tGot 200 response"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.118",
        "@message": "[INFO]\t2025-09-10T16:44:54.118Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.boto3\tdynamodb.GetItem:\t… with a response body of length 69445 starting in b'{\"Item\":{\"expiration\":{\"N\":\"1757522730\"},\"sources\":{\"S\":\"[\\\\\"17f3d434-a25d-4808-9335-f3d9513f4418\\\\\", \\\\\"d549b819-669e-4870-836d-5a07d8f036c4\\\\\", \\\\\"bbe3aa2e-42b0-44db-83c8-a4b9ec8c31ab\\\\\", \\\\\"15daebd5-7c0d-4f02-9c2f-2125a8df363e\\\\\", \\\\\"cc4e4db2-f4cf-4564-92b0-5bda1ea6f069\\\\\", \\\\\"f1a6793e-36f0-431e-b22e-4841d4ac3a0f\\\\\", \\\\\"297dae18-186a-4afc-8cd2-08bf407f79cb\\\\\", \\\\\"0bc55305-9ca2-4cc7-8c04-2862104b4729\\\\\", \\\\\"9f6422b4-08e1-4a37-8910-d846a134149b\\\\\", \\\\\"ff2e5d3c-e8d2-4cea-953f-9a53d244b098\\\\\", \\\\\"1e0e3a10-a591-437d-aa5b-9922ebd02420\\\\\", \\\\\"6a364007-1552-48d9-902d-db2d5e471bef\\\\\", \\\\\"9633c3ad-4a48-435f-9700-d398a535e4f1\\\\\", \\\\\"99274108-6bf3-46c8-8d94-1231b5774628\\\\\", \\\\\"f57ff40f-c923-4356-a0f7-807c82b7063a\\\\\", \\\\\"cfb04b2c-1ef5-43d4-b10c-cc5d7ce7cee8\\\\\", \\\\\"fdb999d0-e6f5-40f7-9d56-1b6138fabf59\\\\\", \\\\\"d31df398-b23f-48ea-a2f3-1ec5bc3537f1\\\\\", \\\\\"c3f66226-3f1f-4fe5-b654-0bb12d62467c\\\\\", \\\\\"26622890-b33f-4c7d-a91e-83e75bda6106\\\\\", \\\\\"9b6a56de-7408-4983-83c3-d3b7689362b1\\\\\", \\\\\"3e8e9a1d-f249-41e7-8a01-ef289ab42241\\\\\", \\\\\"3c3406e5-70e7-4e6a-83bc-a4eddaf9a15d\\\\...'"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.121",
        "@message": "[ERROR]	2025-09-10T16:44:54.119Z	9cba99a5-e83d-4a96-8369-0c550c1e5f12	azul-service-prod	Caught exception for path /index/{entity_type}
Traceback (most recent call last):
  File "/var/task/chalice/app.py", line 1922, in _get_view_function_response
    response = view_function(**function_args)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/app.py", line 1004, in repository_search
    return app.repository_controller.search(catalog=app.catalog,
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/repository_controller.py", line 109, in search
    response = self.service.search(catalog=catalog,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/repository_service.py", line 123, in search
    response = self._search(catalog=catalog,
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/repository_service.py", line 248, in _search
    request = chain.prepare_request(request)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/elasticsearch_service.py", line 135, in prepare_request
    request = self.inner.prepare_request(request)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/elasticsearch_service.py", line 135, in prepare_request
    request = self.inner.prepare_request(request)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/elasticsearch_service.py", line 135, in prepare_request
    request = self.inner.prepare_request(request)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  [Previous line repeated 2 more times]
  File "/var/task/azul/service/elasticsearch_service.py", line 182, in prepare_request
    query = self.prepare_query()
            ^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/elasticsearch_service.py", line 231, in prepare_query
    for field_path, relation_and_values in self.prepared_filters.items():
                                           ^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/caching.py", line 203, in __get__
    value = obj.__dict__[self.fget.__name__] = self.fget(obj)
                                               ^^^^^^^^^^^^^^
  File "/var/task/azul/service/elasticsearch_service.py", line 196, in prepared_filters
    return self._translate_filters(filters_json)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/service/elasticsearch_service.py", line 222, in _translate_filters
    values = field_type.filter(relation, values)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/field.py", line 189, in filter
    return list(map(self.to_index, map(self.from_api, values)))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/field.py", line 144, in from_api
    assert isinstance(value, self.native_types), (value, self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: ({}, <azul.indexer.field.NullableString object at 0x7f66ffb6fad0>)"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.122",
        "@message": "[INFO]\t2025-09-10T16:44:54.121Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.chalice\tReturning 500 response with headers {\"headers\": {\"Access-Control-Allow-Origin\": \"*\", \"Access-Control-Allow-Headers\": \"Authorization,Content-Type,X-Amz-Date,X-Amz-Security-Token,X-Api-Key\", \"Content-Security-Policy\": \"default-src 'self';img-src 'self' data:;script-src 'self';style-src 'self';frame-ancestors 'none';form-action 'self'\", \"Referrer-Policy\": \"strict-origin-when-cross-origin\", \"Strict-Transport-Security\": \"max-age=63072000; includeSubDomains; preload\", \"X-Content-Type-Options\": \"nosniff\", \"X-Frame-Options\": \"DENY\", \"X-XSS-Protection\": \"1; mode=block\", \"Cache-Control\": \"no-store\"}}."
    },
    {
        "@timestamp": "2025-09-10 16:44:54.122",
        "@message": "[INFO]\t2025-09-10T16:44:54.122Z\t9cba99a5-e83d-4a96-8369-0c550c1e5f12\tazul.chalice\t… with a response body of length 133 being {\"Code\": \"InternalServerError\", \"Message\": \"An internal server error occurred.\", \"RequestId\": \"9cba99a5-e83d-4a96-8369-0c550c1e5f12\"}"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.123",
        "@message": "END RequestId: 9cba99a5-e83d-4a96-8369-0c550c1e5f12\n"
    },
    {
        "@timestamp": "2025-09-10 16:44:54.123",
        "@message": "REPORT RequestId: 9cba99a5-e83d-4a96-8369-0c550c1e5f12\tDuration: 13.86 ms\tBilled Duration: 14 ms\tMemory Size: 2048 MB\tMax Memory Used: 265 MB\t\n"
    }
]
```

This was run before the merge of #7453 

## Dev reproduction

(Filter used is `{"projectId":{"is":[{}]}}`)

```
$ curl -s 'https://service.dev.singlecell.gi.ucsc.edu/index/files?filters=%7B%22projectId%22%3A%7B%22is%22%3A%5B%7B%7D%5D%7D%7D'
{"Code":"InternalServerError","Message":"An internal server error occurred.","RequestId":"43913bd9-1941-42d4-82d1-4f7ee39db902"}
```

## Cloudwatch logs (/aws/lambda/azul-service-dev):

```
[ERROR] 2025-10-03T00:39:42.961Z    43913bd9-1941-42d4-82d1-4f7ee39db902    azul-service-dev    Caught exception for path /index/{entity_type}
Traceback (most recent call last):
...
  File "/var/task/azul/indexer/field.py", line 189, in filter
    return list(map(self.to_index, map(self.from_api, values)))
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/var/task/azul/indexer/field.py", line 144, in from_api
    assert isinstance(value, self.native_types), (value, self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
AssertionError: ({}, <azul.indexer.field.NullableString object at 0x7f9a7323cef0>)
```

--- a/lambdas/service/app.py
+++ b/lambdas/service/app.py
@@ -483,21 +483,17 @@ def oauth2_redirect():
                     body=html)
 
 
-def validate_repository_search(entity_type: EntityType,
-                               params: Mapping[str, str],
-                               **validators):
-    validate_params(params, **{
-        'catalog': validate_catalog,
-        'filters': validate_filters,
-        'order': validate_order,
-        'search_after': partial(validate_json_param, 'search_after'),
-        'search_after_uid': str,
-        'search_before': partial(validate_json_param, 'search_before'),
-        'search_before_uid': str,
-        'size': partial(validate_size, entity_type),
-        'sort': validate_field,
-        **validators
-    })
+def validate_repository_search(entity_type: EntityType, params: Mapping[str, str]):
+    validate_params(params,
+                    catalog=validate_catalog,
+                    filters=validate_filters,
+                    order=validate_order,
+                    search_after=partial(validate_json_param, 'search_after'),
+                    search_after_uid=str,
+                    search_before=partial(validate_json_param, 'search_before'),
+                    search_before_uid=str,
+                    size=partial(validate_size, entity_type),
+                    sort=validate_field)
 
 
 def validate_entity_type(entity_type: str):

--- a/lambdas/service/app.py
+++ b/lambdas/service/app.py
@@ -483,19 +483,6 @@ def oauth2_redirect():
                     body=html)
 
 
-def validate_repository_search(entity_type: EntityType, params: Mapping[str, str]):
-    validate_params(params,
-                    catalog=validate_catalog,
-                    filters=validate_filters,
-                    order=validate_order,
-                    search_after=partial(validate_json_param, 'search_after'),
-                    search_after_uid=str,
-                    search_before=partial(validate_json_param, 'search_before'),
-                    search_before_uid=str,
-                    size=partial(validate_size, entity_type),
-                    sort=validate_field)
-
-
 def validate_entity_type(entity_type: str):
     entity_types = app.metadata_plugin.exposed_indices.keys()
     if entity_type not in entity_types:
@@ -990,7 +977,16 @@ def repository_search(entity_type: str, entity_id: str | None = None) -> JSON:
     request = app.current_request
     query_params = request.query_params or {}
     _hoist_parameters(query_params, request)
-    validate_repository_search(entity_type, query_params)
+    validate_params(query_params,
+                    catalog=validate_catalog,
+                    filters=validate_filters,
+                    order=validate_order,
+                    search_after=partial(validate_json_param, 'search_after'),
+                    search_after_uid=str,
+                    search_before=partial(validate_json_param, 'search_before'),
+                    search_before_uid=str,
+                    size=partial(validate_size, entity_type),
+                    sort=validate_field)
     validate_entity_type(entity_type)
     return app.repository_controller.search(catalog=app.catalog,
                                             entity_type=entity_type,

--- a/src/azul/service/__init__.py
+++ b/src/azul/service/__init__.py
@@ -1,6 +1,3 @@
-from collections.abc import (
-    Mapping,
-)
 import logging
 from typing import (
     Protocol,
@@ -44,8 +41,8 @@ type FilterRangeEnd = int | float | str
 # currently represent the mutual exclusivity of the operators. We could, as a
 # union of singleton TypeDict subclasses, but PyCharm doesn't support that.
 #
-FilterOperator = TypedDict(
-    'FilterOperator',
+FilterJSON = TypedDict(
+    'FilterJSON',
     {
         'is': list[PrimitiveJSON | FlatJSON],
         'is_not': list[PrimitiveJSON | FlatJSON],
@@ -56,7 +53,7 @@ FilterOperator = TypedDict(
     total=False
 )
 
-type FiltersJSON = Mapping[FieldName, FilterOperator]
+type FiltersJSON = dict[FieldName, FilterJSON]
 
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True)

--- a/src/azul/service/__init__.py
+++ b/src/azul/service/__init__.py
@@ -37,6 +37,7 @@ log = logging.getLogger(__name__)
 # supports serializing them by default.
 #
 type FilterRange = list[int] | list[float] | list[str]
+type FilterRangeEnd = int | float | str
 
 # `is` is a reserved keyword so we can't use the class-based syntax for
 # TypedDict, but have to use the constructor-based one instead. We don't
@@ -49,7 +50,7 @@ FilterOperator = TypedDict(
         'is': list[PrimitiveJSON | FlatJSON],
         'is_not': list[PrimitiveJSON | FlatJSON],
         'intersects': list[FilterRange],
-        'contains': list[FilterRange | int | float | str],
+        'contains': list[FilterRange | FilterRangeEnd],
         'within': list[FilterRange],
     },
     total=False

--- a/src/azul/service/__init__.py
+++ b/src/azul/service/__init__.py
@@ -1,17 +1,34 @@
+from collections import (
+    defaultdict,
+)
+from functools import (
+    partial,
+)
+from itertools import (
+    chain,
+)
+import json
 import logging
 from typing import (
     Protocol,
     Self,
     TypedDict,
+    cast,
+    get_args,
 )
 
 import attr
 from chalice import (
     ForbiddenError,
 )
+from more_itertools import (
+    one,
+    only,
+)
 
 from azul import (
     CatalogName,
+    R,
     mutable_furl,
 )
 from azul.json import (
@@ -25,6 +42,7 @@ from azul.types import (
     FlatJSON,
     JSON,
     PrimitiveJSON,
+    reify,
 )
 
 log = logging.getLogger(__name__)
@@ -54,6 +72,250 @@ FilterJSON = TypedDict(
 )
 
 type FiltersJSON = dict[FieldName, FilterJSON]
+
+_filter_operators = FilterJSON.__optional_keys__
+_simple_filter_value_types = reify(PrimitiveJSON)
+_dict_filter_value_types = reify(get_args(FlatJSON.__value__)[1])
+assert _simple_filter_value_types == _dict_filter_value_types
+_filter_range_end_types = reify(FilterRangeEnd)
+
+
+def parse_filters(raw_filters: str | None) -> FiltersJSON:
+    """
+    Deserialize, validate and normalize the given string form of the `filters`
+    request parameter. The aim of normalization is to eliminate any
+    insignificant differences so that serializing the value returned from calls
+    to this method with semantically equivalent and valid arguments yields
+    exactly the same JSON string. Two valid arguments are considered
+    semantically equivalent if they match the same subset of all possible
+    documents.
+
+    >>> parse_filters(None)
+    {}
+
+    >>> parse_filters('{}')
+    {}
+
+    >>> parse_filters('{"x":{"is":[null]}}')
+    {'x': {'is': [None]}}
+
+    Values are sorted.
+
+    >>> parse_filters('{"x":{"is":[2,1,null]}}')
+    {'x': {'is': [None, 1, 2]}}
+
+    The entries in value dictionaries are sorted by key.
+
+    >>> parse_filters('{"x":{"is":[{"b":2,"a":1}]}}')
+    {'x': {'is': [{'a': 1, 'b': 2}]}}
+
+    Value dictionaries are sorted by their values, in order of the key. If two
+    dictionaries have equal values at the first key, the value at the second key
+    is used as a tie breaker and so on.
+
+    >>> parse_filters('{"x":{"is":[{"b":2,"a":1},{"a":0,"b":3},{"b":null,"a":1}]}}')
+    {'x': {'is': [{'a': 0, 'b': 3}, {'a': 1, 'b': None}, {'a': 1, 'b': 2}]}}
+
+    Ranges are sorted by start and end value.
+
+    >>> parse_filters('{"x":{"within":[[3,4],[1,2],[1,1]]}}')
+    {'x': {'within': [[1, 1], [1, 2], [3, 4]]}}
+
+    Overall, filters are sorted by field name.
+
+    >>> parse_filters('{"y":{"within":[[1,2]]},"x":{"is":[4, 3]}}')
+    {'x': {'is': [3, 4]}, 'y': {'within': [[1, 2]]}}
+
+    >>> parse_filters('[]')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Filters must be an object')
+
+    >>> parse_filters('{"":42}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Empty field name')
+
+    >>> parse_filters('{"x":42}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Filter must be an object', 'x')
+
+    >>> parse_filters('{"x":{}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Need exactly one filter per field', 'x')
+
+    >>> parse_filters('{"x":{"is":[1],"contains":[1]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Need exactly one filter per field', 'x')
+
+    >>> parse_filters('{"x":{"foo":[2]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Invalid operator', 'x', 'foo')
+
+    >>> parse_filters('{"x":{"is":[]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Need at least one value', 'x')
+
+    >>> parse_filters('{"x":{"is":[1,1]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Duplicate values', 'x')
+
+    >>> parse_filters('{"x":{"within":[1]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Value does not match operator', 'x', <class 'int'>, 'within')
+
+    >>> parse_filters('{"x":{"is":[42,4.1]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Inconsistent value types', 'x')
+
+    >>> parse_filters('{"x":{"within":[[1]]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Range must be list of length 2', 'x')
+
+    >>> parse_filters('{"x":{"within":[[2,1]]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Range is inverted', 'x')
+
+    >>> parse_filters('{"x":{"within":[[1,2],["",""]]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Inconsistent range ends', 'x')
+
+    >>> parse_filters('{"x":{"within":[[1,1.1],[2,2.2]]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Inconsistent range ends', 'x')
+
+    >>> parse_filters('{"x":{"within":[[false,true]]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Invalid range end', 'x')
+
+    >>> parse_filters('{"x":{"within":[{}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Value does not match operator', 'x', <class 'dict'>, 'within')
+
+    >>> parse_filters('{"x":{"within":[[1,2],[1,2]]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Duplicate ranges', 'x')
+
+    >>> parse_filters('{"x":{"is":[{}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Empty object', 'x')
+
+    >>> parse_filters('{"x":{"is":[{"":1}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Empty property name', 'x')
+
+    >>> parse_filters('{"x":{"is":[{"y":1},{"z":2}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Inconsistent property names', 'x')
+
+    >>> parse_filters('{"x":{"is":[{"y":1,"z":[]}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Invalid property value', 'x')
+
+    >>> parse_filters('{"x":{"is":[{"y":1,"z":2},{"y":"","z":3}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Inconsistent property values', 'x')
+
+    >>> parse_filters('{"x":{"is":[{"a":1,"b":2},{"b":2,"a":1}]}}')
+    Traceback (most recent call last):
+        ...
+    AssertionError: R('Duplicate objects', 'x')
+    """
+    if raw_filters is None:
+        return {}
+    else:
+        filters = json.loads(raw_filters)
+        assert type(filters) is dict, R('Filters must be an object')
+        for field, filter in filters.items():
+            assert len(field) > 0, R('Empty field name')
+            assert type(filter) is dict, R('Filter must be an object', field)
+            assert len(filter) == 1, R('Need exactly one filter per field', field)
+            operator, values = one(filter.items())
+            assert operator in _filter_operators, R('Invalid operator', field, operator)
+            assert len(values) > 0, R('Need at least one value', field)
+            num_value_types = set(map(type, values))
+            num_value_types.discard(type(None))
+            assert len(num_value_types) < 2, R('Inconsistent value types', field)
+            value_type = only(num_value_types)
+            mismatch = R('Value does not match operator', field, value_type, operator)
+            if value_type is None:
+                assert operator in {'is'}, mismatch
+            elif value_type in _simple_filter_value_types:
+                assert operator in {'is', 'contains'}, mismatch
+                assert len(set(values)) == len(values), R('Duplicate values', field)
+            elif value_type is list:
+                assert operator in {'contains', 'within', 'intersects'}, mismatch
+                for range in values:
+                    assert len(range) == 2, R('Range must be list of length 2', field)
+                    assert range[0] <= range[1], R('Range is inverted', field)
+                assert len(set(map(tuple, values))) == len(values), R('Duplicate ranges', field)
+                end_types = set(chain.from_iterable(map(partial(map, type), values)))
+                assert len(end_types) == 1, R('Inconsistent range ends', field)
+                end_type = one(end_types)
+                assert end_type in _filter_range_end_types, R('Invalid range end', field)
+            elif value_type is dict:
+                assert operator == 'is', mismatch
+                key_sets = set(map(frozenset, map(dict.keys, values)))
+                assert len(key_sets) == 1, R('Inconsistent property names', field)
+                keys = one(key_sets)
+                assert len(keys) > 0, R('Empty object', field)
+                assert '' not in keys, R('Empty property name', field)
+                value_types_by_key = defaultdict(set)
+                for value in values:
+                    for k, v in value.items():
+                        assert type(v) in _dict_filter_value_types, R('Invalid property value', field)
+                        if v is not None:
+                            value_types_by_key[k].add(type(v))
+                num_value_types = set(map(len, value_types_by_key.values()))
+                assert num_value_types == {1}, R('Inconsistent property values', field)
+                # Sort each value dictionary in place by key (and value, but key
+                # is already unique). This makes sorting the values and checking
+                # their uniqueness easier.
+                for value in values:
+                    sorted_value = dict(sorted(value.items()))
+                    value.clear()
+                    value.update(sorted_value)
+                unique_values = set(map(tuple, map(dict.items, values)))
+                assert len(unique_values) == len(values), R('Duplicate objects', field)
+            else:
+                assert False, R('Invalid value', field)
+
+            def key(v):
+                if v is None:
+                    return False, v
+                elif type(v) is dict:
+                    # The entries in the dict are alteady sorted by key, the
+                    # values are primitive so we just need to handle None values
+                    # and "freeze" the iterable of entries.
+                    return True, tuple((k, key(v)) for k, v in v.items())
+                else:
+                    return True, v
+
+            values.sort(key=key)
+
+        filters = {k: v for k, v in sorted(filters.items())}
+
+        return cast(FiltersJSON, filters)
 
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True)

--- a/src/azul/service/__init__.py
+++ b/src/azul/service/__init__.py
@@ -1,6 +1,5 @@
 from collections.abc import (
     Mapping,
-    Sequence,
 )
 import logging
 from typing import (
@@ -32,10 +31,11 @@ from azul.types import (
 
 log = logging.getLogger(__name__)
 
-# We can't express that these are actually pairs. We could, using tuples, but
-# those are not JSON, generally speaking, even though the `json` module supports
-# serializing them by default.
-FilterRange = Sequence[int] | Sequence[float] | Sequence[str]
+# We can't express that these are actually pairs, i.e. lists of length 2. We
+# could, using tuples, but those are not JSON, even though the `json` module
+# supports serializing them by default.
+#
+FilterRange = list[int] | list[float] | list[str]
 
 # `is` is a reserved keyword so we can't use the class-based syntax for
 # TypedDict, but have to use the constructor-based one instead. We don't
@@ -47,9 +47,9 @@ FilterOperator = TypedDict(
     {
         'is': list[PrimitiveJSON | FlatJSON],
         'is_not': list[PrimitiveJSON | FlatJSON],
-        'intersects': Sequence[FilterRange],
-        'contains': Sequence[FilterRange | int | float | str],
-        'within': Sequence[FilterRange],
+        'intersects': list[FilterRange],
+        'contains': list[FilterRange | int | float | str],
+        'within': list[FilterRange],
     },
     total=False
 )

--- a/src/azul/service/__init__.py
+++ b/src/azul/service/__init__.py
@@ -21,6 +21,7 @@ from azul.json import (
     copy_json,
 )
 from azul.plugins import (
+    FieldName,
     MetadataPlugin,
 )
 from azul.types import (
@@ -54,7 +55,7 @@ FilterOperator = TypedDict(
     total=False
 )
 
-FiltersJSON = Mapping[str, FilterOperator]
+FiltersJSON = Mapping[FieldName, FilterOperator]
 
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True)

--- a/src/azul/service/__init__.py
+++ b/src/azul/service/__init__.py
@@ -36,7 +36,7 @@ log = logging.getLogger(__name__)
 # could, using tuples, but those are not JSON, even though the `json` module
 # supports serializing them by default.
 #
-FilterRange = list[int] | list[float] | list[str]
+type FilterRange = list[int] | list[float] | list[str]
 
 # `is` is a reserved keyword so we can't use the class-based syntax for
 # TypedDict, but have to use the constructor-based one instead. We don't
@@ -55,7 +55,7 @@ FilterOperator = TypedDict(
     total=False
 )
 
-FiltersJSON = Mapping[FieldName, FilterOperator]
+type FiltersJSON = Mapping[FieldName, FilterOperator]
 
 
 @attr.s(auto_attribs=True, kw_only=True, frozen=True)

--- a/src/azul/service/app_controller.py
+++ b/src/azul/service/app_controller.py
@@ -1,4 +1,3 @@
-import json
 from typing import (
     Any,
     Callable,
@@ -8,6 +7,7 @@ from typing import (
 import attr
 from chalice import (
     BadRequestError as BRE,
+    BadRequestError,
     NotFoundError,
 )
 
@@ -22,6 +22,7 @@ from azul.chalice import (
 from azul.service import (
     FileUrlFunc,
     FiltersJSON,
+    parse_filters,
 )
 from azul.strings import (
     pluralize,
@@ -33,14 +34,13 @@ class ServiceAppController(AppController):
     file_url_func: FileUrlFunc
 
     def _parse_filters(self, filters: str | None) -> FiltersJSON:
-        """
-        Parses a string with Azul filters in JSON syntax. Handles default cases
-        where filters are None or '{}'.
-        """
-        if filters is None:
-            return {}
-        else:
-            return json.loads(filters)
+        try:
+            return parse_filters(filters)
+        except AssertionError as e:
+            if R.caused(e):
+                raise R.propagate(e, BadRequestError)
+            else:
+                raise
 
 
 def validate_catalog(catalog):

--- a/src/azul/service/manifest_service.py
+++ b/src/azul/service/manifest_service.py
@@ -110,10 +110,6 @@ from azul.indexer.field import (
 from azul.json import (
     copy_json,
 )
-from azul.json_freeze import (
-    freeze,
-    sort_frozen,
-)
 from azul.plugins import (
     ColumnMapping,
     DocumentSlice,
@@ -984,7 +980,9 @@ class ManifestGenerator(metaclass=ABCMeta):
         different return values.
         """
         git_commit = config.lambda_git_status['commit']
-        filter_string = repr(sort_frozen(freeze(self.filters.explicit)))
+        # The explicit filters are already normalized so we don't to do anything
+        # special to desensitize the hash to insignificat differences
+        filter_string = json.dumps(self.filters.explicit)
         content_hash = str(self.manifest_content_hash)
         catalog = self.catalog
         format = self.format()

--- a/test/service/test_filters.py
+++ b/test/service/test_filters.py
@@ -13,7 +13,7 @@ from azul.plugins import (
     SpecialFields,
 )
 from azul.service import (
-    FilterOperator,
+    FilterJSON,
     Filters,
     FiltersJSON,
 )
@@ -66,7 +66,7 @@ class TestFilterReification(AzulTestCase):
         return filters.reify(plugin=self.plugin, limit_access=limit_access)
 
     def _test_filters(self,
-                      expected_source_id_filter: Optional[FilterOperator],
+                      expected_source_id_filter: Optional[FilterJSON],
                       *,
                       limit_access: bool,
                       explicit_sources: Optional[list[str]] = None,

--- a/test/service/test_manifest.py
+++ b/test/service/test_manifest.py
@@ -923,11 +923,11 @@ class TestManifests(DCP1ManifestTestCase):
                     # a pair of deterministically derived v5 UUIDs.
                     (
                         {'project': {'is': ['Single of human pancreas', 'Mouse Melanoma']}},
-                        'hca-manifest-20d97863-d8cf-54f3-8575-0f9593d3d7ef.4bc67e84-4873-591f-b524-a5fe4ec215eb'
+                        'hca-manifest-89bc9973-de91-5fc4-9c6a-8c1f547d45c6.4bc67e84-4873-591f-b524-a5fe4ec215eb'
                     ),
                     (
                         {},
-                        'hca-manifest-c3cf398e-1927-5aae-ba2a-81d8d1800b2d.4bc67e84-4873-591f-b524-a5fe4ec215eb'
+                        'hca-manifest-832a257c-5540-567b-bcb6-260d2e374508.4bc67e84-4873-591f-b524-a5fe4ec215eb'
                     )
                 ]:
                     with self.subTest(filters=filters, format=format):

--- a/test/service/test_manifest_async.py
+++ b/test/service/test_manifest_async.py
@@ -36,8 +36,8 @@ import requests
 from app_test_case import (
     LocalAppTestCase,
 )
-from azul.collections import (
-    deep_dict_merge,
+from azul.json import (
+    copy_json,
 )
 from azul.logging import (
     configure_test_logging,
@@ -214,7 +214,7 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
         for format, fetch in product([ManifestFormat.compact, ManifestFormat.curl],
                                      [True, False]):
             with self.subTest(format=format, fetch=fetch):
-                filters = {'organ': {'is': ['lymph node']}, 'fileFormat': {'is': ['txt']}}
+                filters = {'fileFormat': {'is': ['txt']}, 'organ': {'is': ['heart', 'lung']}}
                 filters = Filters(explicit=filters, source_ids={self.source.id})
                 params = {
                     'catalog': self.catalog,
@@ -471,7 +471,7 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                         sign_manifest_key.return_value = signed_manifest_key
                     equivalent_url = initial_url.copy()
                     equivalent_filters = json.loads(equivalent_url.args['filters'])
-                    equivalent_filters = dict(reversed(equivalent_filters.items()))
+                    equivalent_filters['organ']['is'].reverse()
                     equivalent_url.args['filters'] = json.dumps(equivalent_filters)
                     url = self._request('PUT', equivalent_url, expect=302)
                     self.assertEqual(final_url, url)
@@ -489,10 +489,8 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                 def modified_put_after_expiration():
                     nonlocal url, state, token_url, equivalent_input
                     get_cached_manifest.side_effect = not_found
-                    equivalent_input = deep_dict_merge(
-                        {'filters': {'explicit': equivalent_filters}},
-                        input
-                    )
+                    equivalent_input = copy_json(input)
+                    equivalent_input['filters']['explicit'] = equivalent_filters
                     iterations.append(equivalent_input)
                     mock_start_generation()
                     url = self._request('PUT', equivalent_url, expect=301)

--- a/test/service/test_manifest_async.py
+++ b/test/service/test_manifest_async.py
@@ -36,9 +36,6 @@ import requests
 from app_test_case import (
     LocalAppTestCase,
 )
-from azul.json import (
-    copy_json,
-)
 from azul.logging import (
     configure_test_logging,
 )
@@ -313,7 +310,6 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                 final_url: furl
                 equivalent_url: furl
                 equivalent_filters: FiltersJSON
-                equivalent_input: ManifestGenerationState
 
                 iterations: list[JSON] = []
 
@@ -475,7 +471,7 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                     equivalent_url.args['filters'] = json.dumps(equivalent_filters)
                     url = self._request('PUT', equivalent_url, expect=302)
                     self.assertEqual(final_url, url)
-                    assert_get_cached_manifest(filters.update(equivalent_filters))
+                    assert_get_cached_manifest(filters)
 
                 modified_put()
 
@@ -487,11 +483,9 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                 #
                 @reset
                 def modified_put_after_expiration():
-                    nonlocal url, state, token_url, equivalent_input
+                    nonlocal url, state, token_url
                     get_cached_manifest.side_effect = not_found
-                    equivalent_input = copy_json(input)
-                    equivalent_input['filters']['explicit'] = equivalent_filters
-                    iterations.append(equivalent_input)
+                    iterations.append(input)
                     mock_start_generation()
                     url = self._request('PUT', equivalent_url, expect=301)
                     assert_get_cached_manifest()
@@ -499,7 +493,7 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                     assert_get_cached_manifest()
                     assert_start_generation()
                     token_url = url
-                    state = equivalent_input
+                    state = input
 
                 modified_put_after_expiration()
                 get_token_while_running()
@@ -516,7 +510,7 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                     state = self.app_module.generate_manifest(state, None)
                     get_cached_manifest_with_key.side_effect = not_found
                     previous_iteration = len(iterations)
-                    iterations.append(equivalent_input)
+                    iterations.append(input)
                     mock_start_generation(start=previous_iteration,
                                           describe=previous_iteration - 1)
                     url = self._request('GET', url, expect=301)
@@ -526,7 +520,7 @@ class TestManifestController(DCP1TestCase, LocalAppTestCase):
                     assert_start_generation(start=previous_iteration,
                                             describe=previous_iteration - 1)
                     token_url = url
-                    state = equivalent_input
+                    state = input
 
                 get_stale_token_when_done()
                 get_token_while_running()

